### PR TITLE
Chado export v2

### PR DESCRIPTION
**Pull request by deepakunni3**
This PR enables the ability to export Apollo annotations to a clean Chado data source.

### CODE REVIEW

File: grails-app/domain/org/bbop/apollo/Genome.groovy [hunk 1]
```diff
@@ -1,16 +0,0 @@
-package org.bbop.apollo
```
Comment (nathandunn) [inline grails-app/domain/org/bbop/apollo/Genome.groovy]: thanks for removing

File: grails-app/domain/org/gmod/chado/Synonym.groovy [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package org.gmod.chado
+
+class Synonym {
+
+    String name
+    String synonymSgml
+//    Serializable searchableSynonymSgml
```
Comment (nathandunn) [inline grails-app/domain/org/gmod/chado/Synonym.groovy]: probably can get rid of this

File: grails-app/services/org/bbop/apollo/ChadoHandlerService.groovy [hunk 3]
```diff
@@ -0,0 +1,1053 @@
...
+ * all feature relationships in feature_relationship table must be SO relationship types
+ * Level 2: Graph - all features relationships between a feature of type X and Y must correspond to relationship of
+ * that type in SO.
+ *
+ * Relevant Chado modules:
+ * Chado General Module
+ * Chado CV Module
+ * Chado Organism Module
+ * Chado Sequence Module
+ * Chado Publication Module
```
Comment (nathandunn) [inline grails-app/services/org/bbop/apollo/ChadoHandlerService.groovy]: Nice notes . . thanks.

File: grails-app/services/org/bbop/apollo/ChadoHandlerService.groovy [hunk 4]
```diff
@@ -0,0 +1,1053 @@
...
+                cvtermsMap.put(it.name, it)
+            }
+        }
+        else {
+            log.warn "No ontologies loaded into the chado database"
+            return null
+        }
+
+        startTime = System.currentTimeMillis()
+        ArrayList<org.gmod.chado.Db> existingChadoDbs = org.gmod.chado.Db.all
```
Comment (nathandunn) [inline grails-app/services/org/bbop/apollo/ChadoHandlerService.groovy]: Probably should be "List" not ArrayList, but not a big deal.

File: sample-docker-apollo-config.groovy [hunk 5]
```diff
@@ -1,3 +1,5 @@
+// TODO: get docker working as well . . .
```
Comment (nathandunn) [inline sample-docker-apollo-config.groovy]: can we create a github issue out of this?

File: src/gwt/org/bbop/apollo/gwt/client/ExportPanel.java [hunk 6]
```diff
@@ -48,24 +59,33 @@ public ExportPanel(OrganismInfo organismInfo, String type, Boolean exportAll, Li
...
         buttonGroup.setDataToggle(Toggle.BUTTONS);
         if (type.equals("FASTA")) {
             buttonGroup.add(genomicRadioButton);
             buttonGroup.add(cdnaRadioButton);
             buttonGroup.add(cdsRadioButton);
             buttonGroup.add(peptideRadioButton);
-        } else if (type.equals("GFF3")) {
+        }
+        else
+        if (type.equals("GFF3")) {
```
Comment (nathandunn) [inline src/gwt/org/bbop/apollo/gwt/client/ExportPanel.java]: We should change this to a FeatureStringEnum (or likewise) at some point

**Comment by nathandunn** [comment]
@deepakunni3  I'm not seeing any red flags in the code . .  now for the testing .. . .

**Comment by nathandunn** [comment]
@deepakunni3  I got the tests to pass (I see 3).  Probably just needed to have it cached.  

The only big problem I can see (still testing) is that we should not have the Chado button (or should have it grayed out) if there is not Chado export option.  Probably the easiest is to grab the config when we get the appState in AnnotatorService.getAppState().     That is when running w/o chado.   Next step is to run with chado so still testing.

**Comment by nathandunn** [comment]
FYI, I'm pushing any of my changes here:  https://github.com/nathandunn/Apollo/commits/deepakunni3-chado_export_v2

**Comment by nathandunn** [comment]
When I run this:

createdb apollo-chado-test
gunzip -c chado-schema-with-ontologies.sql.gz | psql -U ndunn -h localhost -d apollo-chado-test

and I run the integration tests I get 3 errors:

# First Error
```
junit.framework.AssertionFailedError: Condition not satisfied:

lines[2].split("\t")[2]=="gene"
|    |   |          |  |
|    |   |          |  false
|    |   |          |  6 differences (40% similarity)
|    |   |          |  (pseudo)gene
|    |   |          |  (------)gene
|    |   |          pseudogene
|    |   [Group1.10, ., pseudogene, 433519, 437436, ., +, ., Name=GB40815-RAa;date_creation=2016-03-11;ID=d5e4d993-261f-4ee8-bfe4-f1931ed3b318;date_last_modified=2016-03-11]
|    Group1.10	.	pseudogene	433519	437436	.	+	.	Name=GB40815-RAa;date_creation=2016-03-11;ID=d5e4d993-261f-4ee8-bfe4-f1931ed3b318;date_last_modified=2016-03-11
[##gff-version 3, ##sequence-region Group1.10 1 1405242, Group1.10	.	pseudogene	433519	437436	.	+	.	Name=GB40815-RAa;date_creation=2016-03-11;ID=d5e4d993-261f-4ee8-bfe4-f1931ed3b318;date_last_modified=2016-03-11, Group1.10	.	transcript	433519	437436	.	+	.	Name=GB40815-RAa-00001;date_creation=2016-03-11;Parent=d5e4d993-261f-4ee8-bfe4-f1931ed3b318;ID=d90ddf75-fb81-40ee-a753-dae822eab35a;date_last_modified=2016-03-11, Group1.10	.	exon	433519	433570	.	+	.	Name=32cd2626-7df9-4754-b22c-d1ec143900f2-exon;Parent=d90ddf75-fb81-40ee-a753-dae822eab35a;ID=32cd2626-7df9-4754-b22c-d1ec143900f2, Group1.10	.	exon	437425	437436	.	+	.	Name=1b15a6dc-60de-4185-a8f9-96bf12171413-exon;Parent=d90ddf75-fb81-40ee-a753-dae822eab35a;ID=1b15a6dc-60de-4185-a8f9-96bf12171413, Group1.10	.	exon	436577	436641	.	+	.	Name=2cf98c3c-c181-49af-b5af-e4879299916c-exon;Parent=d90ddf75-fb81-40ee-a753-dae822eab35a;ID=2cf98c3c-c181-49af-b5af-e4879299916c, ###, Group1.10	.	gene	1216825	1235616	.	+	.	Name=GB40856-RA;date_creation=2016-03-11;ID=a3bb1f7e-506f-466b-bdac-677dd6394373;date_last_modified=2016-03-11, Group1.10	.	mRNA	1216825	1235616	.	+	.	Name=GB40856-RA-00001;date_creation=2016-03-11;Parent=a3bb1f7e-506f-466b-bdac-677dd6394373;ID=c7a7a3f1-453b-468a-a997-535254306e8e;date_last_modified=2016-03-11, Group1.10	.	CDS	1216825	1216850	.	+	0	Name=1c0d883b-a768-4bb4-a81a-1bfcddb2f468-CDS;Parent=c7a7a3f1-453b-468a-a997-535254306e8e;ID=1c0d883b-a768-4bb4-a81a-1bfcddb2f468, Group1.10	.	CDS	1224677	1224823	.	+	1	Name=1c0d883b-a768-4bb4-a81a-1bfcddb2f468-CDS;Parent=c7a7a3f1-453b-468a-a997-535254306e8e;ID=1c0d883b-a768-4bb4-a81a-1bfcddb2f468, Group1.10	.	CDS	1228683	1228825	.	+	1	Name=1c0d883b-a768-4bb4-a81a-1bfcddb2f468-CDS;Parent=c7a7a3f1-453b-468a-a997-535254306e8e;ID=1c0d883b-a768-4bb4-a81a-1bfcddb2f468, Group1.10	.	CDS	1235238	1235396	.	+	2	Name=1c0d883b-a768-4bb4-a81a-1bfcddb2f468-CDS;Parent=c7a7a3f1-453b-468a-a997-535254306e8e;ID=1c0d883b-a768-4bb4-a81a-1bfcddb2f468, Group1.10	.	CDS	1235488	1235534	.	+	2	Name=1c0d883b-a768-4bb4-a81a-1bfcddb2f468-CDS;Parent=c7a7a3f1-453b-468a-a997-535254306e8e;ID=1c0d883b-a768-4bb4-a81a-1bfcddb2f468, Group1.10	.	exon	1216825	1216850	.	+	.	Name=95bbfcdf-d77b-4c0d-99ce-3f598b7d04e8-exon;Parent=c7a7a3f1-453b-468a-a997-535254306e8e;ID=95bbfcdf-d77b-4c0d-99ce-3f598b7d04e8, Group1.10	.	exon	1228683	1228825	.	+	.	Name=bb66eb10-ab54-4982-aa93-ee77a8c252c9-exon;Parent=c7a7a3f1-453b-468a-a997-535254306e8e;ID=bb66eb10-ab54-4982-aa93-ee77a8c252c9, Group1.10	.	exon	1235238	1235396	.	+	.	Name=33fa1ade-7abb-457f-8c34-da3462290cae-exon;Parent=c7a7a3f1-453b-468a-a997-535254306e8e;ID=33fa1ade-7abb-457f-8c34-da3462290cae, Group1.10	.	exon	1224677	1224823	.	+	.	Name=f6d04238-2d82-4eca-b509-4b9aac524352-exon;Parent=c7a7a3f1-453b-468a-a997-535254306e8e;ID=f6d04238-2d82-4eca-b509-4b9aac524352, Group1.10	.	exon	1235488	1235616	.	+	.	Name=5815c3f7-85bb-405b-bc15-249269d4b65a-exon;Parent=c7a7a3f1-453b-468a-a997-535254306e8e;ID=5815c3f7-85bb-405b-bc15-249269d4b65a, ###, Group1.10	.	insertion	208500	208500	.	+	.	residues=GGG;Name=d5906634-03a4-45aa-a8b1-1f2e3ffab43c-insertion;date_creation=2016-03-11;ID=d5906634-03a4-45aa-a8b1-1f2e3ffab43c;justification=Sanger sequencing;date_last_modified=2016-03-11, ###, Group1.10	.	repeat_region	414370	414600	.	.	.	Name=GB40814-RAa;date_creation=2016-03-11;ID=0aa11542-f9d9-4f02-a4ae-521325503853;date_last_modified=2016-03-11, ###]

	at org.bbop.apollo.Gff3HandlerServiceIntegrationSpec.write a GFF3 of a simple gene model(Gff3HandlerServiceIntegrationSpec.groovy:67)
System output
value of FP to add: Sanger sequencing justification
```

# Second Error

junit.framework.AssertionFailedError: Condition not satisfied:

org.gmod.chado.Organism.count == 1
                        |     |
                        0     false

	at org.bbop.apollo.ChadoHandlerServiceIntegrationSpec.test CHADO export for standard annotations(ChadoHandlerServiceIntegrationSpec.groovy:117)


# Third Error

```
Condition not satisfied: org.gmod.chado.Feature.count > 0 | | 0 false
junit.framework.AssertionFailedError: Condition not satisfied:

org.gmod.chado.Feature.count > 0
                       |     |
                       0     false

	at org.bbop.apollo.ChadoHandlerServiceIntegrationSpec.test CHADO export for annotations with additional information(ChadoHandlerServiceIntegrationSpec.groovy:237)
System output
value of FP to add: 256.7 score
value of FP to add: 9302 score
value of FP to add: 11.2 score
value of FP to add: 42@ score
value of FP to add: Comment for gene null
value of FP to add: Comment for transcript isoform null
```

Going to test actual data running . . .

**Comment by nathandunn** [comment]
When I try to export annotations (off of something roughly equivalent to master I get):

screen shot 2016-03-11 at 4 08 28 pm

Should probably be "OK" instead of "cancel" in both places, but when I added species and organism:

screen shot 2016-03-11 at 4 10 04 pm

Woot!

**Comment by nathandunn** [comment]
Yeah, we maybe shouldn't provide the "select" and "single" option if we really only support a bulk load, but its okay for now.

**Comment by nathandunn** [comment]
Should probably gray out or hide "CHADO" option if not available . . . .but you do catch this on export, which is good.  Pretty minor.

**Comment by nathandunn** [comment]
Export works! 

In Bootstrap.groovy (which loads at startup) . . we need to load ontologies.  I'm wondering if you can check for the datasource there.  

The multiple export is weird, but we can talk about it tomorrow.

**Comment by deepakunni3** [comment]
Based on discussion with Stephen Ficklin, our current Chado export is different in the following three ways:

**1. Store publications in the Chado Publication module, making use of `pub`, `feature_pub` and other related tables.**

Apollo handles publications differently where it adds PMIDs as `dbxref` of a feature. In a Chado export, to create a Chado publication entry it needs a lot more information apart from PMID. We can still go ahead and create entries for the `pub` table but there will be no additional metadata that describes the publication.

**2. Stores GO annotations for features via `feature_cvterm` table.**

Apollo handles GO annotations differently where it adds GO IDs as `dbxref` of a feature. In a Chado export, the limitation is that we cannot create an entry for `feature_cvterm` table without having an associated publication that serves as provenance for the GO annotation.

**3. Uses `synonym` table for storing symbol of a feature.**

Apollo has symbol as an attribute of a feature. In a Chado export, although we can create an entry for the `synonym` table, we cannot link a `synonym` with a `feature`, via `feature_synonym` table, because this requires an associated publication that serves as provenance.


@monicacecilia The above three points are open for further discussion.

**Comment by monicacecilia** [comment]
@deepakunni3 Noted. thank you! :bowtie:  Sent PM to you & @nathandunn

**Comment by nathandunn** [comment]
Quick comments (since I'm off most of next week):
1 - if there is a way to hide the "Chado" export that would be good, but not imperative.
2 - if ontologies are loaded it seems to hang (maybe the reason for it being "slow")
3 - its still pretty slow.  I'm not sure if its the writing or the reading.   The more objects that can be collected via a join (versus iteration), the better (see the GFF3 export, e.g.).   For doing inserts . . . if its possible to do insertions in bulk and then add relationships later, that is better, but not imperative.   Also, its not a particularly intensive process, so you could toss it into the background (if all else fails) to "finish up later", if they want it to work in the background.

**Comment by nathandunn** [comment]
So, it is hanging . . . it creates the schema correctly on loadup, but its not loading the ontologies automatically via bootstrap:

```
                def proc = "scripts/load_chado.sh ${username} ${dbName}".execute()
                proc.waitForProcessOutput(outputBuffer, errorBuffer)
                println "output: ${outputBuffer}"
                println "error : ${errorBuffer}"
                log.debug "Error: ${errorBuffer.toString()}"```

The speed is probably fine for now.  It looks like its being added twice possibly:

```
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
error : ERROR:  language "plpgsql" already exists
ERROR:  relation "cvtermpath" already exists
ERROR:  relation "feature" already exists
ERROR:  relation "featureloc" already exists
ERROR:  relation "cvterm" already exists
ERROR:  relation "db" already exists
ERROR:  relation "dbxref" already exists
ERROR:  relation "feature_cvterm" already ex
```

**Comment by nathandunn** [comment]
I think the issue is that Grails is actually loading the chado database according to the preprogrammed schema and then this script loads the database again with the ontologies, potentially causing it to not load since the schema was already loaded.   May just need to split the script up.

**Comment by nathandunn** [comment]
@deepakunni3  Is this ready for me to look at?

**Comment by deepakunni3** [comment]
I have a few more improvements to push. One I have that done, it should be ready.

**Comment by nathandunn** [comment]
Perfect.  I’ll just wait for it then.

**Comment by deepakunni3** [comment]
@nathandunn This is ready for review again.

File: scripts/load_chado_schema.sh [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    esac
+done
+
+if [ -z "${PG_USER}" ] || [ -z "${DB}" ] || [ -z "${CHADO_SCHEMA}" ] ; then
+    usage
+fi
+
+check_config
+
+psql -U $PG_USER -h $HOST -p $PORT -c "CREATE DATABASE \"$DB\""
```
Comment (nathandunn) [inline scripts/load_chado_schema.sh]: @deepakunni3   This is the only thing that seemed to be creating problems.      I added an echo statement and I get this:

```
./scripts/load_chado_schema.sh -u nathandunn -d apollo-chado -h localhost -r -s 5432 -s chado-schema-with-ontologies.sql.gz
psql -U nathandunn -h localhost -p 5432 -c CREATE DATABASE "apollo-chado"
psql: FATAL:  database "nathandunn" does not exist
Cannot create database 'apollo-chado' due to improper connection parameters, lack of privileges or non-existent user 'nathandunn'.
```

However, I know this works this way:

```
createdb -U nathandunn apollo-chado 
```

**Comment by cmdcolin** [comment]
This is maybe not a pre-requisite for this PR, but I think that this should add to the "data_adapters" configuration element for what types of options are available. Then, the chado exporter can be added to the list of data_adapters too. Furthermore, the sidebar should interpret the data_adapters config too, which would make options for exporting more flexible

**Comment by nathandunn** [comment]
Added issue to #953.

**Comment by nathandunn** [comment]
Just a quick note @deepakunni3   Most of the other stuff looked good (other than the database script).   I got it up and running without chado and the chado box was gray, which is correct.  The documentation and script looked good as well.  

Its possible that it is just that might environment is screwed up, but I think that it general the "psql" command expects a database of some sort, even if you are using it create another one.

**Comment by deepakunni3** [comment]
@nathandunn Thanks for reviewing the PR. 

I have seen that psql error before, not in the context of this PR, and I can reproduce it. It seems PostgreSQL expects a database of the same name as the username. This is because the `psql` command tries to connect to a database before running any statement on behalf of the user. Instead we can change this behavior by explicitly specifying `template1` database.

I'll make the changes after testing.

**Comment by nathandunn** [comment]
@deepakunni3   did you want me to continue testing before you change it?   Sorry, wasn't sure what you meant.

**Comment by deepakunni3** [comment]
@nathandunn The fix for the psql issue has been pushed. I was implementing the ability to poll the server for export status. Maybe that can be worked on separately instead of including it to be part of this PR. If you feel the same then you can go ahead and test again :)

**Comment by nathandunn** [comment]
I think that separate is fine . . . I'll give it another test.

File: grails-app/controllers/org/bbop/apollo/IOServiceController.groovy [hunk 8]
```diff
@@ -126,7 +129,7 @@ class IOServiceController extends AbstractApolloController {
                 } else {
                     gff3HandlerService.writeFeaturesToText(outputFile.path, features, grailsApplication.config.apollo.gff3.source as String)
                 }
-            } else if (typeOfExport == "FASTA") {
+            } else if (typeOfExport == FeatureStringEnum.TYPE_FASTA.getValue()) {
```
Comment (nathandunn) [inline grails-app/controllers/org/bbop/apollo/IOServiceController.groovy]: nice . .  thanks for fixing all of these

File: grails-app/conf/BootStrap.groovy [hunk 9]
```diff
@@ -11,6 +11,7 @@ import org.bbop.apollo.Transcript
 import org.bbop.apollo.User
 import org.bbop.apollo.UserService
 import org.bbop.apollo.sequence.SequenceTranslationHandler
+import org.codehaus.groovy.grails.commons.GrailsApplication
```
Comment (nathandunn) [inline grails-app/conf/BootStrap.groovy]: I think you can probably get rid of a few more unused statements here.

**Comment by nathandunn** [comment]
If I run this:

```
./scripts/load_chado_schema.sh -u nathandunn -d apollo-chado -s chado-schema-with-ontologies.sql.gz -r
```

And I go to export I get this:

screen shot 2016-03-31 at 11 37 43 am

```
| Error Caused by: org.postgresql.util.PSQLException: ERROR: permission denied for relation feature
| Error 	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2161)
| Error 	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1890)
| Error 	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:255)
| Error 	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:559)
| Error 	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:417)
| Error 	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeQuery(AbstractJdbc2Statement.java:302)
| Error 	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:80)
| Error 	at org.hibernate.loader.Loader.getResultSet(Loader.java:2065)
| Error 	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1862)
| Error 	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1838)
| Error 	at org.hibernate.loader.Loader.doQuery(Loader.java:909)
| Error 	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:354)
| Error 	at org.hibernate.loader.Loader.doList(Loader.java:2553)
| Error 	at org.hibernate.loader.Loader.doList(Loader.java:2539)
| Error 	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2369)
| Error 	at org.hibernate.loader.Loader.list(Loader.java:2364)
| Error 	at org.hibernate.loader.criteria.CriteriaLoader.list(CriteriaLoader.java:126)
| Error 	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1682)
| Error 	at org.hibernate.internal.CriteriaImpl.list(CriteriaImpl.java:380)
| Error 	at org.hibernate.internal.CriteriaImpl.uniqueResult(CriteriaImpl.java:402)
| Error 	at org.codehaus.groovy.grails.orm.hibernate.HibernateGormStaticApi$3.doInHibernate(HibernateGormStaticApi.groovy:236)
| Error 	at org.codehaus.groovy.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:179)
| Error 	... 156 more
```

**Comment by deepakunni3** [comment]
In `apollo-config.groovy`, is the username for the chado data source `nathandunn`?

**Comment by nathandunn** [comment]
Good call . . retesting . . .

**Comment by deepakunni3** [comment]
Although, this is good. I should add code for catching that exception.

**Comment by nathandunn** [comment]
that seems to fixed it . . but then I ran out of memory . . .

**Comment by nathandunn** [comment]
@deepakunni3   I keep getting out of memory errors and it takes a LONG time . . .

**Comment by nathandunn** [comment]
much faster when not importing the sequences!

**Comment by monicacecilia** [comment]
Merged??!! Great news! :+1:

**Pull request merged**